### PR TITLE
Fix updateProjectFile/resetProjectFile potential problems.

### PR DESCRIPTION
**Pull request by NozomiIto**
While investigating my xcodebuild code 65 error and updatedWDABundleId capability behavior, I found the following problems when I used updatedWDABundleId:

- When I tried simple dummy updatedWDABundleId value like "A" or "B", WebDriverAgent/WebDriverAgent.xcodeproj/project.pbxproj become broken. This is because resetProjectFile method replaced many occurrences of "A" or "B" in project.pbxproj despite these strings are not part of PRODUCT_BUNDLE_IDENTIFIER value.
- When Appium server is killed forcibly and resetProjectFile is not called unfortunately, project.pbxproj file keeps having the user specific bundle ID. When this happened, udpateProjectFile method cannot replace bundle ID again, since now project.pbxproj does not have the "com.facebook.WebDriverAgentRunner" string occurrence..

This PR resolves these problems by using project.pbxproj.old to restore the original project.pbxproj file.

- This restoration is called even if updatedWDABundleId is not specified, since test with updatedWDABundleId capabilities may have broken project.pbxproj.
- I checked the old invalid project.pbxproj.old will be removed when `npm install` for new version appium-xcuitest-driver is called. So we don't need to care about the possibility of the existing invalid project.pbxproj and project.pbxproj.old files.
- I use node.js native `fs.existsSync` instead of `fs.exists` of appium-support, since [fs.exists](https://github.com/appium/appium-support/blob/master/lib/fs.js#L23) implementation calls `this.hasAccess(path)` without async and caused the timing problem on my PC. I think this appium-support implementation also should be fixed, but I hesitate to do so since fs.exists is used for many Appium logic and this fix can affect them..

### CODE REVIEW

File: lib/wda/utils.js [hunk 1]
```diff
@@ -35,12 +42,20 @@ async function updateProjectFile (agentPath, newBundleId) {
...
+ * @param {string} agentPath - Path to the .xcodeproj directory.
+ */
+async function resetProjectFile (agentPath) {
   let projectFilePath = `${agentPath}/${PROJECT_FILE}`;
   try {
-    await replaceInFile(projectFilePath, new RegExp(newBundleId.replace('.', '\.'), 'g'), WDA_RUNNER_BUNDLE_ID);
-    await fs.unlink(`${projectFilePath}.old`);
-    log.debug(`Successfully reset '${projectFilePath}' with bundle id '${WDA_RUNNER_BUNDLE_ID}'`);
+    // restore projectFilePath from .old file
+    // TODO currently avoid using fs.exists since its implementation seems not be able to async/await well
```
Comment (imurchie) [inline lib/wda/utils.js]: What do you mean by this?

File: lib/wda/utils.js [hunk 2]
```diff
@@ -35,12 +42,20 @@ async function updateProjectFile (agentPath, newBundleId) {
...
+    if (nativeFs.existsSync(`${projectFilePath}.old`)) {
```
Comment (imurchie) [inline lib/wda/utils.js]: We can't have sync functions here.

**Comment by NozomiIto** [comment]
@imurchie Thanks for your quick review!
You mean I should use 
```
if (await fs.exists(\${projectFilePath}.old`) {
```
instead of
```
if (nativeFs.existsSync(`${projectFilePath}.old`)) {
```

right?

**Comment by NozomiIto** [comment]
@imurchie Thanks for your quick review!
You mean I should use 
```
if (await fs.exists(\${projectFilePath}.old`) {
```
or
```
if (await fs.exists(\${projectFilePath}.old`) {
```

instead of
```
if (nativeFs.existsSync(`${projectFilePath}.old`)) {
```

right?

**Comment by imurchie** [comment]
Yes.  The sync function is blocking, so should be avoided.

**Comment by imurchie** [comment]
This scenario may also need some consideration with regard to resource locking to handle multiple sessions. Not sure about that. Cc: @mykola-mokhnach

File: lib/wda/utils.js [hunk 3]
```diff
@@ -35,12 +42,20 @@ async function updateProjectFile (agentPath, newBundleId) {
   }
 }
 
-async function resetProjectFile (agentPath, newBundleId) {
+/**
+ * Reset WebDriverAgentRunner project bundle ID to correct state.
+ * @param {string} agentPath - Path to the .xcodeproj directory.
+ */
+async function resetProjectFile (agentPath) {
   let projectFilePath = `${agentPath}/${PROJECT_FILE}`;
```
Comment (mykola-mokhnach) [inline lib/wda/utils.js]: I'd rather implement quick return right here:

```js
if (!await fs.exists(`${projectFilePath}.old`)) {
  return;
}
```

**Comment by mykola-mokhnach** [comment]
Yep, there might be issues with parallel sessions. The following situation is possible:
1. Session 1 patches the project file
2. Session 1 starts building
3. Session 2 resets the file
4. Session 1 uses the reseted file for building.
5. Session 2 patches the file and is built properly.

We can use two approaches here to solve the problem: 
- Combine patching and building under a single lock, so each session prepares its own env (simpler approach)
- Add a separate static primitive to make sure the second session does not rewrite the config for the previous one (complex approach)

File: lib/wda/utils.js (hunk 3 above)
Comment (NozomiIto) [reply lib/wda/utils.js]: I see.

**Comment by NozomiIto** [comment]
> Yep, there might be issues with parallel sessions. The following situation is possible:

Thanks. I try to implement it.

> Yes. The sync function is blocking, so should be avoided.

Yeah, I understand.
But I already tried `await fs.exists()` but I found `fs.exists` (this is appium-support function) sometimes returns true even if the target file has already been removed, and it did not work as I expected..
I think [this line](https://github.com/appium/appium-support/blob/master/lib/fs.js#L23) should be `async exists (path) { return await this.hasAccess(path); }` instead of `exists (path) { return this.hasAccess(path); }`...

**Comment by mykola-mokhnach** [comment]
> I think this line should be async exists (path) { return await this.hasAccess(path); } instead of exists (path) { return this.hasAccess(path); }...

This should not be a problem, since hasAccess is already returning a promise, which means it will still return a promise even if you don't explicitly mark it as async.

Regarding locks and parallel sessions: I've just created another [PR](https://github.com/appium/appium-xcuitest-driver/pull/538), which should simplify the stuff, by applying resource lock on top driver level. So, you don't have to think about synchronization after it is merged.

**Comment by NozomiIto** [comment]
> Regarding locks and parallel sessions: I've just created another PR
Thank you so much!
OK, I just think about fs.exists problem.

**Comment by NozomiIto** [comment]
Fixed the problems above.

> This should not be a problem, since hasAccess is already returning a promise, which means it will still return a promise even if you don't explicitly mark it as async.

Thanks! My understanding of fs.exists was wrong. 

I use fs.mv instead of fs.copyFile and fs.unlink combination, since this combination does not work well.(I'm afraid this is fs.copyFile problem, but I don't investigate the detail)

File: lib/wda/utils.js [hunk 4]
```diff
@@ -35,11 +41,18 @@ async function updateProjectFile (agentPath, newBundleId) {
...
   try {
-    await replaceInFile(projectFilePath, new RegExp(newBundleId.replace('.', '\.'), 'g'), WDA_RUNNER_BUNDLE_ID);
-    await fs.unlink(`${projectFilePath}.old`);
+    // restore projectFilePath from .old file
+    if (!fs.exists(`${projectFilePath}.old`)) {
```
Comment (mykola-mokhnach) [inline lib/wda/utils.js]: missing await

Comment (NozomiIto) [reply lib/wda/utils.js]: Oh, sorry. fixed.

**Comment by imurchie** [comment]
With the change in resource locking there are conflicts now. Please fix them and we can get this in!

**Comment by NozomiIto** [comment]
Fixed the conflict.

**Comment by imurchie** [comment]
Published in `appium-xcuitest-driver@2.53.0`

**Pull request merged**